selectrum.el

```diff
@@ -1339,6 +1339,105 @@ semantics of `cl-defun'."
            prompt nil keymap nil
            (or history 'minibuffer-history)))))))
 
+(defvar selectrum--completing-read-handler-alist
+  '((selectrum-locate-library
+     :transformer selectrum--locate-library-transformer)
+    (org-tags-completion-function
+     :transformer selectrum--org-tags-completion-transformer
+     :handler selectrum--org-tags-completing-handler)
+    (org-agenda-filter-completion-function
+     ;; This table was mentioned on the org mailing list to have the
+     ;; same problems as org-tags-completion-function. Seems not to be
+     ;; as easy to simply delegate it to `completing-read-multiple'
+     ;; for now just remap it to use default completion.
+     :handler completing-read-default))
+  "Used to handle some special case completion tables.
+
+Some dynamic completion tables are specifically designed for
+`completing-read-default'. To make those tables work under the
+Selectrum paradigm or to fix other problematic tables this list
+contains information how to handle them in
+`selectrum-completing-read'.
+
+The keys are either the name determined by
+`selectrum--get-collection-name' or the completion table
+name (fallback).
+
+The values are plists using the following keys:
+
+:transformer
+
+  A function to transform the completion table. Receives the
+  original completion table and should return a completion table
+  which should be used instead. If passed a completion function
+  the transformer should also return a completion function so the
+  table doesn't get initialized before
+  `selectrum--minibuffer-post-command-hook' runs. This can be
+  useful when one wants to skip the initial computation of the
+  table. With `completing-read-default' the computation happens
+  only after pressing TAB and this allows to mimic this behavior.
+
+:handler
+
+  A function to be used as `completing-read-function' for the
+  table. This handler gets passed the original table or the
+  result of :transformer if specified.")
+
+(defun selectrum--get-collection-name (collection)
+  "Get symbol for completion COLLECTION.
+The symbol is used to identify COLLECTION in
+`selectrum--completing-read-handler-alist'"
+  (cond ((and (memq real-this-command
```

Comment:
Is there any way we can avoid relying on `real-this-command`?

Reply:
I don't know of a better alternative. We could remove that check for `real-this-command` entirely, though. I put it there because I wanted to narrow the scope down to the cases I have tested. Removing duplicates shouldn't break anything but removing the directory entries maybe. This is another example of a table which assumes to be used by the default completion paradigm. The `elc` and directory candidates show only up when there are no other candidates left (in default completion that means you already have inserted the lib name and then press TAB again to see the `elc` and dir entry for that lib in the completion list).

Reply:
Hm. Would it be possible to condition the entire alist only on the value of `this-command` (or `real-this-command`)? That would make it easier for the end user to change, as well as simplify logic.

Reply:
One problem is that  `this-command` and `real-this-command` can be unreliable (#98). That is also the reason that ivy has chosen to let you pass the `:caller` argument for each command. But maybe we could avoid that by tracking commands as proposed in #98. But I think usually dispatching based on completion table is better. It works for all commands which use the same table (regardless of the command which called it). This case is a bit special unfortunately because it's a completion table that gets created by 

```elisp
(apply-partially
   'locate-file-completion-table ...
```

and the arguments passed vary by command.

> That would make it easier for the end user to change, as well as simplify logic

I didn't intend the alist to be changed by users and made it a private variable. Would you prefer to have it public?

Reply:
Okay. This entire situation is somewhat distasteful; I suppose there is no perfect solution. Your approach seems fine.